File: 알고리즘 기본 구현/DisjointSet.cpp
#include <vector>

class DisjointSet
{
private:
    std::vector<int> parent;
public:
    DisjointSet(int count)
    {
        parent.resize(count);
        for (int i = 0; i < count; ++i)
            parent[i] = i;
    }

    int findParent(int x)
    {
        if (x == parent[x])
            return x;
        
        return parent[x] = findParent(parent[x]);
    }

    void unionNodes(int x, int y)
    {
        int xPar = findParent(x);
        int yPar = findParent(y);

        parent[xPar] = yPar;
    }
};

File: 알고리즘 기본 구현/DisjointSet.rs
struct DisjointSet {
    parent: Vec<usize>
}

impl DisjointSet {
    fn new(size: usize) -> DisjointSet {
        let parent = (0..size).collect();
        DisjointSet { parent }
    }

    fn find_parent(&mut self, x: usize) -> usize {
        let p = self.parent.get(x).expect("DisjointSet: x is out of range");
        if *p == x {
            x
        } else {
            let new_parent = self.find_parent(p.clone());
            self.parent[x] = new_parent;
            new_parent 
        }
    }

    fn union_nodes(&mut self, x: usize, y: usize) {
        let x_par = self.find_parent(x);
        let y_par = self.find_parent(y);
        self.parent[x_par] = y_par;
    }
}